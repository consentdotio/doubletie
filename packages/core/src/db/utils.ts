/**
 * Database utility functions
 */
import { StandardSchemaV1 } from '@standard-schema/spec';
import type {
	EntitySchemaDefinition,
	FieldValueType,
	SchemaField,
} from '../schema/schema.types';
import { getAdapter } from './adapters';
import type { TableDefinition } from './adapters/adapter';

/**
 * Helper function to generate a table definition for an entity
 * @param entity The entity definition
 * @param adapterType The database adapter type to use
 * @returns A table definition
 */
export function generateTableDefinition(
	entity: EntitySchemaDefinition,
	adapterType = 'sqlite'
): TableDefinition {
	const adapter = getAdapter(adapterType);
	if (!adapter) {
		throw new Error(`Adapter not found for type: ${adapterType}`);
	}
	return adapter.generateTableDefinition(entity);
}

/**
 * Helper function to generate SQL for creating a table from an entity
 * @param entity The entity definition
 * @param adapterType The database adapter type to use
 * @returns SQL string for creating the table
 */
export function generateSQLForEntity(
	entity: EntitySchemaDefinition,
	adapterType = 'sqlite'
): string {
	const adapter = getAdapter(adapterType);
	const tableDef = adapter.generateTableDefinition(entity);
	return adapter.generateCreateTableSQL(tableDef);
}

/**
 * Helper function to transform a value to database format
 * @param field The field definition
 * @param value The value to transform
 * @param adapterType The database adapter type to use
 * @returns The transformed value
 */
export function toDatabaseValue<TField>(
	field: TField,
	value: SchemaField<
		string,
		FieldValueType,
		StandardSchemaV1<unknown, unknown>
	>,
	adapterType = 'sqlite'
): unknown {
	const adapter = getAdapter(adapterType);
	return adapter.toDatabase(field, value);
}
/**
 * Helper function to transform a value from database format
 * @param field The field definition
 * @param dbValue The database value to transform
 * @param adapterType The database adapter type to use
 * @returns The transformed value
 */
export function fromDatabaseValue<TField>(
	field: TField,
	dbValue: SchemaField<
		string,
		FieldValueType,
		StandardSchemaV1<unknown, unknown>
	>,
	adapterType = 'sqlite'
): any {
	const adapter = getAdapter(adapterType);
	return adapter.fromDatabase(field, dbValue);
}
