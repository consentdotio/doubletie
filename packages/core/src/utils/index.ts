/**
 * Utility module exports
 * @module utils
 */

// Export type guards
export * from './type-guards';
