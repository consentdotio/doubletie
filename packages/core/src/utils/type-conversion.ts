import type { StandardSchemaV1 } from '@standard-schema/spec';
import type { EntityFromDefinition } from '../entity/entity.types';
import type {
	EntitySchemaDefinition,
	FieldValueType,
} from '../schema/schema.types';

/**
 * Convert EntityFromDefinition to EntitySchemaDefinition
 * This is needed for compatibility with database adapters
 * @param entity - The entity to convert
 * @returns The converted entity schema definition
 */
export function toEntitySchemaDefinition<T extends EntityFromDefinition>(
	entity: T
): EntitySchemaDefinition {
	// Validate the entity has the required properties
	if (!entity || typeof entity !== 'object') {
		throw new Error('Invalid entity: must be an object');
	}

	if (!entity.name || typeof entity.name !== 'string') {
		throw new Error('Invalid entity: must have a name property of type string');
	}

	if (!entity.fields || typeof entity.fields !== 'object') {
		throw new Error(
			'Invalid entity: must have a fields property of type object'
		);
	}

	// Create the schema definition with proper type checking
	const result: EntitySchemaDefinition = {
		name: entity.name,
		fields: entity.fields,
		description:
			typeof entity.description === 'string' || entity.description === undefined
				? entity.description
				: undefined,
	};

	// Add optional properties with type checking
	if (entity.prefix && typeof entity.prefix === 'string') {
		result.prefix = entity.prefix;
	}

	if (
		entity.config &&
		typeof entity.config === 'object' &&
		entity.config !== null
	) {
		result.config = entity.config as Record<string, FieldValueType>;
	}

	if (typeof entity.order === 'number') {
		result.order = entity.order;
	}

	if (
		entity.validator &&
		typeof entity.validator === 'object' &&
		'~standard' in entity.validator &&
		typeof entity.validator['~standard'] === 'object' &&
		typeof entity.validator['~standard'].validate === 'function'
	) {
		result.validator = entity.validator;
	}

	return result;
}
