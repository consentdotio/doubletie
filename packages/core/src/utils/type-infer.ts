/**
 * Type inference utility for doubletie core
 *
 * This file contains utility types that help with type inference.
 */

import type { StandardSchemaV1 } from '@standard-schema/spec';
import type {
	EntityFromDefinition,
	EntityStructure,
	EntityWithRelationships,
} from '../entity/entity.types';
import type { EntityFieldReference } from '../entity/relationship.types';
import type { FieldValueType, SchemaField } from '../schema/schema.types';
import type {
	FieldType,
	FieldTypeToValueType,
} from '../validation/type-conversion';

/**
 * Infer the entity type from a relationship reference
 */
export type InferEntityFromReference<T extends EntityFieldReference> =
	T extends EntityFieldReference<infer S, infer E, infer F> ? E : never;

/**
 * Infer the field type from a relationship reference
 */
export type InferFieldFromReference<T extends EntityFieldReference> =
	T extends EntityFieldReference<infer S, infer E, infer F> ? F : never;

/**
 * Infer the source entity type from a relationship reference
 */
export type InferSourceFromReference<T extends EntityFieldReference> =
	T extends EntityFieldReference<infer S, infer E, infer F> ? S : never;

/**
 * Infer the value type from a schema field
 */
export type InferValueFromField<F extends SchemaField> = F extends SchemaField<
	infer T,
	infer V
>
	? V extends FieldValueType
		? FieldTypeToValueType<T & FieldType>
		: V
	: never;

/**
 * Extract all field values from an entity as a record type
 */
export type EntityRecord<T extends EntityStructure> = {
	[K in keyof T['fields']]: InferValueFromField<T['fields'][K]>;
};

/**
 * Extract relationships from an entity with relationships
 */
export type EntityRelationships<
	T extends EntityWithRelationships<
		EntityStructure,
		Record<string, EntityFieldReference>
	>,
> = T extends EntityWithRelationships<EntityStructure, infer R>
	? R
	: Record<string, never>;

/**
 * Get the input type for an entity (for validation)
 */
export type EntityInput<T extends EntityFromDefinition> =
	T extends EntityFromDefinition<infer S, infer V>
		? V extends StandardSchemaV1
			? StandardSchemaV1.InferInput<V>
			: EntityRecord<T>
		: Record<string, FieldValueType>;

/**
 * Extract a type from a relationship field
 */
export type RelatedEntityType<
	TEntity extends EntityWithRelationships<
		EntityStructure,
		Record<string, EntityFieldReference>
	>,
	TFieldName extends keyof EntityRelationships<TEntity>,
> = InferEntityFromReference<EntityRelationships<TEntity>[TFieldName & string]>;

/**
 * Check if a type is an array type
 */
export type IsArray<T> = T extends Array<unknown> ? true : false;

/**
 * Create a deep partial type that makes all properties optional recursively
 */
export type DeepPartial<T> = T extends object
	? { [P in keyof T]?: DeepPartial<T[P]> }
	: T;

/**
 * Create a type with non-nullable fields
 */
export type NonNullableFields<T> = {
	[P in keyof T]: NonNullable<T[P]>;
};

/**
 * Makes properties in T with types that match K non-nullable
 */
export type RequireFields<T, K extends keyof T> = T & {
	[P in K]-?: NonNullable<T[P]>;
};

/**
 * Extract the type of properties with a specific field type
 */
export type FieldsByType<
	T extends EntityStructure,
	TFieldType extends FieldType,
> = {
	[K in keyof T['fields']]: T['fields'][K] extends SchemaField<
		TFieldType,
		infer V
	>
		? InferValueFromField<T['fields'][K]>
		: never;
};

/**
 * Creates a record type with all required fields from an entity
 */
export type RequiredEntityRecord<T extends EntityStructure> = {
	[K in keyof T['fields'] as T['fields'][K]['required'] extends true
		? K
		: never]-?: InferValueFromField<T['fields'][K]>;
} & {
	[K in keyof T['fields'] as T['fields'][K]['required'] extends true
		? never
		: K]?: InferValueFromField<T['fields'][K]>;
};
