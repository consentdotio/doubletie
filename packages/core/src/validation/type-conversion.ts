import type { SchemaField } from '../schema/schema.types';

/**
 * Common field types used in the schema system
 */
export type FieldType =
	| 'string'
	| 'number'
	| 'boolean'
	| 'date'
	| 'timestamp'
	| 'uuid'
	| 'id'
	| 'incremental_id'
	| 'json'
	| 'array'
	| string;

/**
 * Map field types to their JavaScript type
 */
export type FieldTypeToValueType<T extends FieldType> = T extends
	| 'string'
	| 'uuid'
	| 'id'
	? string
	: T extends 'number' | 'incremental_id'
		? number
		: T extends 'boolean'
			? boolean
			: T extends 'date' | 'timestamp'
				? Date
				: T extends 'json'
					? Record<string, unknown>
					: T extends 'array'
						? Array<unknown>
						: unknown;

/**
 * Helper function to convert values to appropriate types based on field definition
 * @param value The value to convert
 * @param fieldDef The field definition
 * @param fieldName The name of the field
 * @returns The converted value with the appropriate type
 */
export function convertValueToFieldType<
	TFieldType extends FieldType = FieldType,
	TValue = FieldTypeToValueType<TFieldType>,
>(
	value: unknown,
	fieldDef: SchemaField<TFieldType, TValue>,
	fieldName: string
): TValue {
	// Handle null explicitly
	if (value === null) {
		return null as TValue;
	}

	// If the field has a custom input transform, use it
	if (fieldDef.transform?.input) {
		return fieldDef.transform.input(value);
	}

	// Handle conversion based on field type
	const fieldType = fieldDef.type;

	// Timestamp/Date fields
	const timestampPatterns = ['createdAt', 'updatedAt', 'deletedAt'];
	const timestampSuffixes = ['At', '_at'];

	if (
		fieldType === 'date' ||
		fieldType === 'timestamp' ||
		timestampPatterns.includes(fieldName) ||
		timestampSuffixes.some((suffix) => fieldName.endsWith(suffix))
	) {
		// Convert to Date if not already
		return (value instanceof Date ? value : new Date(String(value))) as TValue;
	}

	// ID fields - ensure they're strings (except incremental IDs)
	const idPatterns = ['id'];
	const idSuffixes = ['Id', '_id'];

	if (
		idPatterns.includes(fieldName) ||
		fieldDef.primaryKey === true ||
		idSuffixes.some((suffix) => fieldName.endsWith(suffix)) ||
		fieldType === 'uuid' ||
		fieldType === 'id'
	) {
		// Only convert to string if it's not an incremental ID
		if (fieldType !== 'incremental_id' && fieldType !== 'number') {
			return String(value) as TValue;
		}
	}

	// Number fields
	if (fieldType === 'number' || fieldType === 'incremental_id') {
		// Convert to number if not already
		return (typeof value === 'number' ? value : Number(value)) as TValue;
	}

	const isTrueString = (str: string) =>
		['true', 'yes', 'y', '1'].includes(str.toLowerCase());

	const isFalseString = (str: string) =>
		['false', 'no', 'n', '0'].includes(str.toLowerCase());

	// Boolean fields
	if (
		fieldType === 'boolean' ||
		fieldName.startsWith('is') ||
		fieldName.startsWith('has')
	) {
		// Convert to boolean if not already
		if (typeof value === 'boolean') {
			return value as TValue;
		}
		if (typeof value === 'string') {
			if (isTrueString(value)) return true as TValue;
			if (isFalseString(value)) return false as TValue;
		}
		return Boolean(value) as TValue;
	}

	// For all other cases, return as is
	return value as TValue;
}
