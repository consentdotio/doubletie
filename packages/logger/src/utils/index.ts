// Export result logging utilities
export {
	logResult,
	logResultAsync,
	logResult as logError,
	logResultAsync as logErrorAsync,
} from './result';

// Export telemetry utilities
export {
	withLogSpan,
	setDefaultTracer,
	setTelemetryDisabled,
	setDefaultAttributes,
} from './telemetry';
