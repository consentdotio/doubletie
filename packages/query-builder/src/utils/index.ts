/**
 * Utility functions for the ORM
 *
 * @module utils
 */

export { default as isolate } from './isolate.js';
export * from './compose.js';
export * from './json-helper.js';
export * from './model-builder.js';
export * from './type-inference.js';
