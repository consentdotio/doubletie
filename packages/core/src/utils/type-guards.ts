/**
 * Type guard utilities for doubletie core
 *
 * This file contains functions that help with type safety by providing runtime
 * type checking that also informs TypeScript about the types.
 */

import type { StandardSchemaV1 } from '@standard-schema/spec';
import type { DatabaseConfig } from '../config/config.types';
import type { ResolvedEntitySchema } from '../db/merge';
import type { SqlColumnType, SqlDialect, TableDefinition } from '../db/table';
import type { EntityStructure } from '../entity/entity.types';
import type {
	EntityFieldReference,
	JoinTableConfig,
	RelationshipConfig,
	RelationshipType,
} from '../entity/relationship.types';
import type { FieldValueType, SchemaField } from '../schema/schema.types';
import { FieldType } from '../validation/type-conversion';
import type { ValidationError } from '../validation/validate';

// Basic type guards
/**
 * Check if a value is a string
 */
export function isString(value: unknown): value is string {
	return typeof value === 'string';
}

/**
 * Check if a value is a number
 */
export function isNumber(value: unknown): value is number {
	return typeof value === 'number' && !isNaN(value);
}

/**
 * Check if a value is a boolean
 */
export function isBoolean(value: unknown): value is boolean {
	return typeof value === 'boolean';
}

/**
 * Check if a value is a function
 */
export function isFunction(value: unknown): value is Function {
	return typeof value === 'function';
}

/**
 * Check if a value is an array
 */
export function isArray(value: unknown): value is unknown[] {
	return Array.isArray(value);
}

/**
 * Check if a value is a non-null object
 */
export function isObject(value: unknown): value is Record<string, unknown> {
	return typeof value === 'object' && value !== null;
}

/**
 * Check if a value is a Date
 */
export function isDate(value: unknown): value is Date {
	return value instanceof Date;
}

/**
 * Check if a value is null
 */
export function isNull(value: unknown): value is null {
	return value === null;
}

/**
 * Check if a value is undefined
 */
export function isUndefined(value: unknown): value is undefined {
	return value === undefined;
}

/**
 * Check if a value is a valid field value type
 */
export function isFieldValueType(value: unknown): value is FieldValueType {
	return (
		isString(value) ||
		isNumber(value) ||
		isBoolean(value) ||
		isDate(value) ||
		isObject(value) ||
		isNull(value) ||
		isUndefined(value) ||
		isArray(value)
	);
}

// Entity type guards
/**
 * Check if a value is a valid EntityStructure
 */
export function isEntityStructure(value: unknown): value is EntityStructure {
	return (
		isObject(value) &&
		'name' in value &&
		isString(value.name) &&
		'fields' in value &&
		isObject(value.fields)
	);
}

/**
 * Check if a value is a valid EntitySchemaDefinition with required properties
 */
export function isEntitySchemaDefinition(value: unknown): boolean {
	return (
		isEntityStructure(value) &&
		'name' in value &&
		isString(value.name) &&
		'fields' in value &&
		isObject(value.fields)
	);
}

/**
 * Check if a value is a valid SchemaField
 */
export function isSchemaField<T extends FieldType = FieldType>(
	value: unknown
): value is SchemaField<T> {
	return isObject(value) && 'type' in value && isString(value.type);
}

/**
 * Check if a value is a valid Standard Schema validator
 */
export function isStandardSchema(value: unknown): value is StandardSchemaV1 {
	return (
		isObject(value) &&
		'~standard' in value &&
		isObject(value['~standard']) &&
		'validate' in value['~standard'] &&
		isFunction(value['~standard'].validate)
	);
}

/**
 * Type guard to check if a value is a record of a specific type
 */
export function isRecordOf<T>(
	value: unknown,
	itemTypeGuard: (item: unknown) => item is T
): value is Record<string, T> {
	if (!isObject(value)) {
		return false;
	}

	return Object.values(value).every((item) => itemTypeGuard(item));
}

/**
 * Type guard to check if a value is an array of a specific type
 */
export function isArrayOf<T>(
	value: unknown,
	itemTypeGuard: (item: unknown) => item is T
): value is T[] {
	return isArray(value) && value.every((item) => itemTypeGuard(item));
}

/**
 * Type guard to check if a string is a valid FieldType
 */
export function isFieldType(value: string): value is FieldType {
	return (
		[
			'string',
			'number',
			'boolean',
			'date',
			'timestamp',
			'uuid',
			'id',
			'incremental_id',
			'json',
			'array',
		].includes(value) || isString(value)
	);
}

/**
 * Type guard to check if a string is a valid RelationshipType
 */
export function isRelationshipType(value: string): value is RelationshipType {
	return ['oneToOne', 'oneToMany', 'manyToOne', 'manyToMany'].includes(value);
}

/**
 * Type guard to check if an object is a valid JoinTableConfig
 */
export function isJoinTableConfig(value: unknown): value is JoinTableConfig {
	if (!isObject(value)) {
		return false;
	}

	return (
		'name' in value &&
		isString(value.name) &&
		'sourceColumn' in value &&
		isString(value.sourceColumn) &&
		'targetColumn' in value &&
		isString(value.targetColumn)
	);
}

/**
 * Type guard to check if an object is a valid RelationshipConfig
 */
export function isRelationshipConfig(
	value: unknown
): value is RelationshipConfig {
	if (!isObject(value)) {
		return false;
	}

	if (
		!('type' in value && isString(value.type) && isRelationshipType(value.type))
	) {
		return false;
	}

	// Different validation based on relationship type
	if (value.type === 'manyToMany') {
		return (
			'joinTable' in value &&
			(isString(value.joinTable) || isJoinTableConfig(value.joinTable))
		);
	} else {
		// For other relationship types, foreignKey is required
		return 'foreignKey' in value && isString(value.foreignKey);
	}
}

/**
 * Type guard to check if a value is an EntityFieldReference
 */
export function isEntityFieldReference(
	value: unknown
): value is EntityFieldReference {
	if (!isObject(value)) {
		return false;
	}

	return (
		'entity' in value &&
		isString(value.entity) &&
		'field' in value &&
		isString(value.field)
	);
}

// Additional helper for ValidationError issues check
export function isValidationIssue(
	item: unknown
): item is { field: string; issues: string } {
	return (
		isObject(item) &&
		'field' in item &&
		isString(item.field) &&
		'issues' in item &&
		isString(item.issues)
	);
}

/**
 * Type guard to check if a value is a ValidationError
 */
export function isValidationError(error: unknown): error is ValidationError {
	return (
		error instanceof Error &&
		'issues' in error &&
		isArrayOf((error as ValidationError).issues, isValidationIssue)
	);
}

/**
 * Type guard to check if a value is a valid DatabaseConfig
 */
export function isDatabaseConfig(value: unknown): value is DatabaseConfig {
	if (!isObject(value)) {
		return false;
	}

	return 'adapter' in value && isString(value.adapter);
}

/**
 * Type guard to check if a value is a valid SQL dialect
 */
export function isSqlDialect(value: unknown): value is SqlDialect {
	return isString(value) && ['postgresql', 'mysql', 'sqlite'].includes(value);
}

/**
 * Type guard to check if a value is a valid SQL column type
 */
export function isSqlColumnType(value: unknown): value is SqlColumnType {
	const validTypes = [
		'VARCHAR',
		'TEXT',
		'INTEGER',
		'BIGINT',
		'DECIMAL',
		'FLOAT',
		'BOOLEAN',
		'DATE',
		'TIMESTAMP',
		'JSON',
		'JSONB',
		'UUID',
		'ARRAY',
	];

	return isString(value) && validTypes.includes(value);
}

/**
 * Type guard to check if a value is a TableDefinition
 */
export function isTableDefinition(value: unknown): value is TableDefinition {
	if (!isObject(value)) {
		return false;
	}

	return (
		'name' in value &&
		isString(value.name) &&
		'columns' in value &&
		isObject(value.columns) &&
		'primaryKey' in value &&
		(isString(value.primaryKey) || isArrayOf(value.primaryKey, isString))
	);
}

/**
 * Type guard to check if a value is a ResolvedEntitySchema
 */
export function isResolvedEntitySchema(
	value: unknown
): value is ResolvedEntitySchema<any> {
	if (!isObject(value)) {
		return false;
	}

	return (
		'entityName' in value &&
		isString(value.entityName) &&
		'entityPrefix' in value &&
		isString(value.entityPrefix) &&
		'fields' in value &&
		isObject(value.fields)
	);
}

/**
 * Assert that a value has a specific type using a type guard
 * @throws Error if the value doesn't match the type guard
 */
export function assertType<T>(
	value: unknown,
	typeGuard: (val: unknown) => val is T,
	errorMessage: string
): asserts value is T {
	if (!typeGuard(value)) {
		throw new Error(errorMessage);
	}
}

/**
 * Assert that a condition is true, with a custom error message
 * @throws Error if the condition is false
 */
export function assertCondition(
	condition: boolean,
	errorMessage: string
): asserts condition {
	if (!condition) {
		throw new Error(errorMessage);
	}
}
