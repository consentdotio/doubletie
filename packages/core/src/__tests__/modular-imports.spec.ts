import { describe, expect, test, vi } from 'vitest';
import { toEntitySchemaDefinition } from '../utils/type-conversion';

// We need to test that specific modules can be imported independently
describe('Modular Imports', () => {
	test('Entity module can be imported independently', async () => {
		// Import only entity-related functionality
		const { defineEntity } = await import('../entity/entity');

		// Verify the import worked
		expect(defineEntity).toBeDefined();
		expect(typeof defineEntity).toBe('function');

		// Test basic functionality
		const simpleEntity = defineEntity({
			name: 'test',
			fields: {},
		});

		expect(simpleEntity).toBeDefined();
		expect(simpleEntity.name).toBe('test');
	});

	test('Schema fields module can be imported independently', async () => {
		// Import only schema field utilities
		const { createdAtField, emailField, stringField, uuidField } = await import(
			'../schema/fields'
		);

		// Verify imports worked
		expect(createdAtField).toBeDefined();
		expect(emailField).toBeDefined();
		expect(stringField).toBeDefined();
		expect(uuidField).toBeDefined();

		// Test basic functionality
		const idField = uuidField();
		const nameField = stringField({ required: true });
		const emailField1 = emailField({ required: true });
		const timestampField = createdAtField();

		expect(idField.type).toBe('string');
		expect(nameField.required).toBe(true);
		expect(emailField1.validator).toBeDefined();
		expect(timestampField.defaultValue).toBeDefined();
	});

	test('Database module can be imported independently', async () => {
		// Mock UUID for consistent snapshots
		vi.mock('uuid', () => ({
			v4: () => '123e4567-e89b-12d3-a456-426614174000',
		}));

		// Import only database-related functionality
		const { generateSQLForEntity, generateTableDefinition } = await import(
			'../db'
		);

		// Import dependencies needed to create an entity
		const { defineEntity } = await import('../entity/entity');
		const { uuidField, stringField, emailField, createdAtField } = await import(
			'../schema/fields'
		);

		// Create a test entity
		const userEntity = defineEntity({
			name: 'user',
			fields: {
				id: uuidField(),
				username: stringField({
					required: true,
					databaseHints: { maxSize: 50, unique: true },
				}),
				email: emailField({ required: true }),
				createdAt: createdAtField(),
			},
		});

		// Verify database functions work
		expect(generateSQLForEntity).toBeDefined();
		expect(generateTableDefinition).toBeDefined();

		const sqliteDef = generateTableDefinition(
			toEntitySchemaDefinition(userEntity),
			'sqlite'
		);

		// Basic verification of table definition
		expect(sqliteDef).toBeDefined();
		expect(sqliteDef.name).toBe('user');

		// Find columns by name
		const idColumn = sqliteDef.columns.find((col) => col.name === 'id');
		const usernameColumn = sqliteDef.columns.find(
			(col) => col.name === 'username'
		);
		const emailColumn = sqliteDef.columns.find((col) => col.name === 'email');

		expect(idColumn).toBeDefined();
		expect(usernameColumn).toBeDefined();
		expect(emailColumn).toBeDefined();

		// Create SQL for the entity
		const sqliteSQL = generateSQLForEntity(
			toEntitySchemaDefinition(userEntity),
			'sqlite'
		);
		const postgresSQL = generateSQLForEntity(
			toEntitySchemaDefinition(userEntity),
			'postgres'
		);

		// Basic SQL verification
		expect(sqliteSQL).toContain('CREATE TABLE');
		expect(sqliteSQL).toContain('user');
		expect(postgresSQL).toContain('CREATE TABLE');
		expect(postgresSQL).toContain('user');

		// Use snapshots for SQL validation
		expect(sqliteSQL).toMatchSnapshot('sqlite-sql');
		expect(postgresSQL).toMatchSnapshot('postgres-sql');
	});

	// Skip validation tests for now since they require more complex setup
	test.skip('Validation module can be imported independently', async () => {
		// This test is skipped until we can properly mock the validation dependencies
	});

	test.skip('A complete example works with modular imports', async () => {
		// This test is skipped until we can properly mock the validation dependencies
	});
});
