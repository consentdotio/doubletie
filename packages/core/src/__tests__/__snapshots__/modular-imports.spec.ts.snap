// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modular Imports > Database module can be imported independently > postgres-sql 1`] = `
"CREATE TABLE IF NOT EXISTS "user" (
  "id" UUID NOT NULL PRIMARY KEY UNIQUE DEFAULT NULL,
  "username" VARCHAR(50) NOT NULL UNIQUE,
  "email" VARCHAR NOT NULL,
  "createdAt" TIMESTAMP NOT NULL DEFAULT NULL
);
"
`;

exports[`Modular Imports > Database module can be imported independently > sqlite-sql 1`] = `
"CREATE TABLE IF NOT EXISTS "user" (
  "id" TEXT PRIMARY KEY NOT NULL UNIQUE DEFAULT NULL,
  "username" TEXT NOT NULL UNIQUE DEFAULT NULL,
  "email" TEXT NOT NULL DEFAULT NULL,
  "createdAt" TEXT NOT NULL DEFAULT NULL
);"
`;
